### PR TITLE
Networking: Add support for multiqueue tap

### LINKED ISSUE #247: Networking - Enable Multi-Queue virtio-net
_From @mcastelino on August 24, 2016 16:10_

Multi-queue virtio-net provides an approach that scales the network performance as the number of vCPUs increases, by allowing them to transfer packets through more than one virtqueue pair at a time.

_Copied from original issue: 01org/cc-oci-runtime#162_

### PR DESCRIPTION
**Pull request by mcastelino**
Networking: Add support for multiqueue tap

Add support for multiqueue tap for the bridge mode.
This improves the multi-q performance of bridge mode.
However macvtap mode is still faster.

Fixes #247

Requires merge of https://github.com/vishvananda/netlink/pull/293 followed by netlink revendoring

### CODE REVIEW

**Comment by jodh-intel** [comment]
aside from the conflict...

lgtm.

File: network.go [hunk 1]
```diff
@@ -523,6 +524,35 @@ func createLink(netHandle *netlink.Handle, name string, expectedLink netlink.Lin
 	return getLinkByName(netHandle, name, expectedLink)
 }
 
+func createMqLink(netHandle *netlink.Handle, name string, expectedLink netlink.Link) (netlink.Link, []*os.File, error) {
```
Comment (sboeuf) [inline network.go]: Instead of adding this function, don't you think we could add a flag for the multiqueue to `createLink()`:
```go
func createLink(netHandle *netlink.Handle, name string, expectedLink netlink.Link, mq bool)
```

File: qemu.go [hunk 2]
```diff
@@ -313,7 +313,7 @@ func (q *qemu) appendSocket(devices []ciaoQemu.Device, socket Socket) []ciaoQemu
 func networkModelToQemuType(model NetInterworkingModel) ciaoQemu.NetDeviceType {
 	switch model {
 	case ModelBridged:
-		return ciaoQemu.TAP
+		return ciaoQemu.MACVTAP
```
Comment (sboeuf) [inline qemu.go]: Why did you change this ?

File: network.go [hunk 3]
```diff
@@ -523,6 +524,35 @@ func createLink(netHandle *netlink.Handle, name string, expectedLink netlink.Lin
...
+		}
+	default:
+		return nil, fds, fmt.Errorf("Unsupported link type %s", expectedLink.Type())
+	}
+
+	if err := netHandle.LinkAdd(newLink); err != nil {
+		return nil, fds, fmt.Errorf("LinkAdd() failed for %s name %s: %s", expectedLink.Type(), name, err)
+	}
+
+	tuntapLink, ok := newLink.(*netlink.Tuntap)
```
[the remark this answers is not in the document]
Comment (mcastelino) [reply network.go]: @sboeuf  Good catch will fix

File: qemu.go (hunk 2 above)
Comment (mcastelino) [reply qemu.go]: @sboeuf The mq link uses the same connection methodology as macvtap. Hence the change. Longer term I should change the name of the ciaoQEMU type should be ciaoQemu.FD. macvtap was too specific as it is just a fd based networking connection.

I will rename/rework the qemu networking code quite a bit when adding support for vhost-user and macvlan and tc based connectivity. Will do the renaming as part of that just to avoid revendoring qemu.

File: network.go (hunk 1 above)
Comment (mcastelino) [reply network.go]: @sboeuf I should have removed the old method completely. It is no longer needed. We will be switching to fd based methods for all networking interfaces as far as possible. This will give us more control on how we setup the interfaces. i.e. QEMU will not be performing any networking setup internally.

File: network.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcastelino) [reply network.go]: @sboeuf fixed

**Comment by mcastelino** [comment]
@chavafg can you look at why the CI is failing. There are two vendorings here. And somehow our CI is not picking them up properly
```
23:48:49 # github.com/clearcontainers/runtime/vendor/github.com/containers/virtcontainers
23:48:49 vendor/github.com/containers/virtcontainers/network.go:517: unknown field 'Queues' in struct literal of type netlink.Tuntap
23:48:49 vendor/github.com/containers/virtcontainers/network.go:518: undefined: netlink.TUNTAP_MULTI_QUEUE_DEFAULTS
23:48:49 vendor/github.com/containers/virtcontainers/network.go:546: tuntapLink.Fds undefined (type *netlink.Tuntap has no field or method Fds)
23:48:49 vendor/github.com/containers/virtcontainers/qemu.go:357: unknown field 'VhostFDs' in struct literal of type qemu.NetDevice
```

**Comment by mcastelino** [comment]
Closing this to reopen it using a different branch as the CI keeps failing due to vendoring.
